Suppress attribute feature form if no fields in layer

nyalldawson opened the pull request:
This change prevents an empty form popup for layers with no fields (eg memory layers). If desired, this behaviour can be overriden by the layer's feature form suppress option, eg if form has python logic which makes it useful even with no fields. 

The change was discussed in https://hub.qgis.org/issues/10775 and set as wontfix at the time, but I believe that by respecting the layer's feature form suppress option we don't lose any functionality by this change. If acceptable I will backport this fix to 2.8 too.

(also includes a tiny change to make the scratch layer editable by default)

nyalldawson committed:
Make new scratch layer editable by default

nyalldawson committed:
Suppress attribute form popup if layer has no fields

Prevents empty form popup for layers with no fields (eg memory layers).
If desired, this behaviour can be overriden by the layer's feature
form suppress option, eg if form has python logic which makes it useful
even with no fields.

(fix #10775)

nyalldawson commented:
@m-kuhn thoughts?

nirvn commented:
@nyalldawson one thing to consider; some time ago, I was discussing with Martin of adding a [ continue input ] button next to the [ ok ] and [ cancel ] buttons to recover from an accidental right click. 

The following scenario drives me _mad_: I have entered over 50 nodes of a complex polygon, I'm halfway there, and bam, I accidently hit the right button on my mouse and have no way to continue. I either accept the feature and add the missing nodes via double clicking on the feature line, or I cancel the feature, which means I have to restart from scratch.

If, and I pray to god, this button is ever added, killing the form for feature-less layers would be problematic.

m-kuhn commented:
I had a discussion some time ago with @3nids and we thought that it would be better to set the "suppress feature form" flag of the layer when creating an attribute-less layer.
This would allow to re-enable it and therefore not interfere with the possibility to have some other fancy logic one may have in the feature form (E.g. somebody will print the coordinates with some information about the vegetation at this location...).

pcav commented:
Sounds good.

nyalldawson commented:
manually merged after discussion with @m-kuhn - given that the implementation allows the form to be shown by setting the suppress behaviour for the layer, this addresses all concerns
